### PR TITLE
Da: network core

### PR DESCRIPTION
**Pull request by danielSanchezQ**
* Sketch up da network libp2p stack
* Implemented first rough version of broadcast protocol

### CODE REVIEW

File: nomos-da/network/subnetworks-assignations/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+use std::collections::HashSet;
+use std::hash::Hash;
+
+pub trait MembershipHandler {
```
Comment (bacv) [inline nomos-da/network/subnetworks-assignations/src/lib.rs:4]: As of now I feel that this trait would better fit inside the `network/core/membership/mod.rs`. The actual functions to calculate the subnetwork idx from the pov key could also live there (assuming `nomos-da` crates are considered as a libraries).

The composition of these parts will happen in `nomos-services`, where we'll be able to provide different adapters (eg. `MockMemebershipAdapter`?

File: nomos-da/network/core/src/broadcast/behaviour.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+use subnetworks_assignations::MembershipHandler;
+
+use super::handler::{BroadcastHandler, DaMessage, HandlerEventToBehaviour};
+
+pub type SubnetworkId = u16;
+
+/// Nomos DA BroadcastEvents to be bubble up to logic layers
+pub enum BroadcastEvent {
+    IncomingMessage { peer_id: PeerId, message: DaMessage },
```
Comment (bacv) [inline nomos-da/network/core/src/broadcast/behaviour.rs:19]: Looks like we won't need packing after all. Seems that libp2p will take care of it. Will update messages crate if true.

**Review by bacv — APPROVED** [review]
Great progress, it would be great to have a simple service as a next step to integrate the messages and quic transport.

**Comment by danielSanchezQ** [comment]
> Great progress, it would be great to have a simple service as a next step to integrate the messages and quic transport.

I think a test instead will be better, shouldn't be too complicated, just spawning 2 "nodes" as tasks and interchange a few random blobs.

File: nomos-da/network/core/src/broadcast/behaviour.rs (hunk 2 above)
Comment (danielSanchezQ) [reply nomos-da/network/core/src/broadcast/behaviour.rs:19]: > Looks like we won't need packing after all. Seems that libp2p will take care of it. Will update messages crate if true.

It makes sense is like that, Thanks for taking a look at it.

File: nomos-da/network/subnetworks-assignations/src/lib.rs (hunk 1 above)
Comment (danielSanchezQ) [reply nomos-da/network/subnetworks-assignations/src/lib.rs:4]: Yeah, the trait would fit there as well. But I think as those algorithms are "simple" but will change in the future they deserved their own crate. Then in the nomos-services we just compose them indeed.

Comment (bacv) [reply nomos-da/network/subnetworks-assignations/src/lib.rs:4]: 🤝

**Review by bacv — APPROVED** [review]

**Comment by danielSanchezQ** [comment]
Unrelated lint. Merging

**Pull request merged**